pkg/hosts/dialer/ssh.go

```diff
@@ -126,95 +130,147 @@ func (d *SSHDialer) Dial(t bool) (*ssh.Client, error) {
 	return ssh.Dial("tcp", d.sshAddress, cfg)
 }
 
-// NewSession returns new session.
-func (d *SSHDialer) NewSession() error {
+func (d *SSHDialer) GetClient() *ssh.Client {
+	return d.conn
+}
+
+func (d *SSHDialer) getUserID() error {
+	if d.uid >= 0 {
+		return nil
+	}
 	session, err := d.conn.NewSession()
 	if err != nil {
 		return err
 	}
-	d.session = session
+	defer func() { _ = session.Close() }()
+
+	output, err := session.Output("id -u")
+	if err != nil {
+		return fmt.Errorf("failed to get current user id from remote host %s, %v", d.sshAddress, err)
+	}
+	// it should return a number with user id if ok
+	d.uid, err = strconv.Atoi(strings.TrimSpace(string(output)))
+	if err != nil {
+		return fmt.Errorf("failed to parse uid output from remote host, output: %s, %v", string(output), err)
+	}
+	return nil
+}
+
+func (d *SSHDialer) Close() error {
+	if d.conn != nil {
+		return d.conn.Close()
+	}
+	for _, shell := range d.shells {
+		shell.Close()
+		delete(d.shells, shell)
+	}
+	return nil
+}
+
+func (d *SSHDialer) wrapCommands(cmd string) string {
+	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(scriptWrapper, cmd)))
+}
+
+func (d *SSHDialer) ExecuteCommands(cmds ...string) (string, error) {
+	if err := d.getUserID(); err != nil {
+		return "", err
+	}
+
+	sudo := ""
+	if d.uid > 0 {
+		sudo = "sudo"
+	}
+
+	session, err := d.conn.NewSession()
+	if err != nil {
+		return "", err
+	}
+	defer session.Close()
+
+	encodedCMD := d.wrapCommands(strings.Join(cmds, "\n"))
+	cmd := fmt.Sprintf("echo \"%s\" | base64 -d | %s sh -", encodedCMD, sudo)
+	d.logger.Debugf("executing cmd: %s", cmd)
+
+	output := bytes.NewBuffer([]byte{})
+	combinedOutput := singleWriter{
+		b: io.MultiWriter(d.logger.Out, output),
```

Comment:
Where do we initialize the mutext lock for singleWriter?

Reply:
Here is copy code from combine output, as we need to set combineOutput to both stdout and stderr.